File: frontend/src/interfaces/Mapset.ts
interface Mapset {
    id: number;
    artist: string;
    title: string;
    covers: { [key: string]: string };
    status: string;
    last_updated: string; // Assuming time.Time is serialized as string in JSON
    user_id: number;
    preview_url: string;
    tags: string;
    mapset_stats: MapsetStats;
    bpm: number;
    creator: string;
    beatmaps: Beatmap[];
}

interface MapsetStats {
    [key: string]: MapsetStatsModel;
}

interface MapsetStatsModel {
    play_count: number;
    favorite_count: number;
    comments_count: number;
}

interface MapsetStatsDataset {
    timestamp: string;
    play_count: number;
    favorite_count: number;
    comments_count: number;
}

File: backend/internal/database/model/mapset.go
package model

import (
	"encoding/json"
	"time"
)

type Mapset struct {
	ID            int
	Artist        string
	Title         string
	Covers        json.RawMessage `gorm:"type:jsonb"` // map[string]string probably
	Status        string
	LastUpdated   time.Time
	UserID        int
	Creator       string
	Language      string
	Genre         string
	PreviewURL    string
	Tags          string
	BPM           float64
	MapsetStats   json.RawMessage `gorm:"type:jsonb"` // MapsetStats struct marshaled as JSON
	LastPlaycount int
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

type MapsetStats map[time.Time]*MapsetStatsModel

type MapsetStatsModel struct {
	Playcount int `json:"play_count"`
	Favorites int `json:"favorite_count"`
	Comments  int `json:"comments_count"`
}
